frame/support/procedural/src/partial_eq_no_bound.rs

```diff
@@ -0,0 +1,124 @@
+		Ok(input) => input,
+		Err(e) => return e.to_compile_error().into(),
+	};
+
+	let name = &input.ident;
+	let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+
+	let impl_ = match input.data {
+		syn::Data::Struct(struct_) => match struct_.fields {
+			syn::Fields::Named(named) => {
+				let fields = named.named.iter()
+					.map(|i| i.ident.as_ref().expect("named fields have ident"))
+					.map(|i| quote::quote_spanned!(i.span() => self.#i == other.#i ));
+
+				quote::quote!( true #( && #fields )* )
+			},
+			syn::Fields::Unnamed(unnamed) => {
+				let fields = unnamed.unnamed.iter().enumerate()
+					.map(|(i, _)| syn::Index::from(i))
+					.map(|i| quote::quote_spanned!(i.span() => self.#i == other.#i ));
+
+				quote::quote!( true #( && #fields )* )
+			},
+			syn::Fields::Unit => {
+				quote::quote!( true )
+			}
+		},
+		syn::Data::Enum(enum_) => {
+			let variants = enum_.variants.iter()
+				.map(|variant| {
+					let ident = &variant.ident;
+					match &variant.fields {
+						syn::Fields::Named(named) => {
+							let names = named.named.iter()
+								.map(|i| i.ident.as_ref().expect("named fields have ident"));
+							let names_bis = names.clone()
+								.map(|i| syn::Ident::new(&format!("{}_bis", i), i.span()));
+
+							let capture = names.clone();
+							let capture_bis = names.clone().zip(names_bis.clone())
+								.map(|(i, i_bis)| quote::quote!(#i: #i_bis));
+							let eq = names.zip(names_bis)
+								.map(|(i, i_bis)| quote::quote_spanned!(i.span() => #i == #i_bis));
+							quote::quote!(
+								(
+									Self::#ident { #( #capture, )* },
+									Self::#ident { #( #capture_bis, )* },
+								) => true #( && #eq )*
+							)
+						},
+						syn::Fields::Unnamed(unnamed) => {
+							let names = unnamed.unnamed.iter().enumerate()
+								.map(|(i, f)| syn::Ident::new(&format!("_{}", i), f.span()));
+							let names_bis = unnamed.unnamed.iter().enumerate()
+								.map(|(i, f)| syn::Ident::new(&format!("_{}_bis", i), f.span()));
+							let eq = names.clone().zip(names_bis.clone())
+								.map(|(i, i_bis)| quote::quote_spanned!(i.span() => #i == #i_bis));
+							quote::quote!(
+								(
+									Self::#ident ( #( #names, )* ),
+									Self::#ident ( #( #names_bis, )* ),
+								) => true #( && #eq )*
+							)
+						},
+						syn::Fields::Unit => quote::quote!( (Self::#ident, Self::#ident) => true ),
+					}
+				});
+
+			let mut different_variants = vec![];
+			for (i, i_variant) in enum_.variants.iter().enumerate() {
+				for (j, j_variant) in enum_.variants.iter().enumerate() {
+					if i != j {
+						let i_ident = &i_variant.ident;
+						let j_ident = &j_variant.ident;
+						different_variants.push(quote::quote!(
+							(Self::#i_ident { .. }, Self::#j_ident { .. }) => false
+						))
+					}
+				}
+			}
+
+			quote::quote!( match (self, other) {
+				#( #variants, )*
+				#( #different_variants, )*
+			})
```

Comment:
```suggestion
			quote::quote!( match (self, other) {
				#( #variants, )*
				_ => false,
			})
```

Isn't this the same and much shorter?

Reply:
I was thinking of a case where another attribute add a variant on the enum.
```
#[derive(PartialEqNoBound)]
#[some_macro_which_add_a_variant_somehow]
pub enum Foo {}
```

(I'm actually not sure in which order they are called)
here I would prefer the macro not to compile than to have a unexpected implementation of partial eq.

Reply:
Good point.

While the attribute macro needs to be called before the derive macros, as they would fail otherwise :D